geode-book/Gemfile

```diff
@@ -17,6 +17,7 @@
 
 source "https://rubygems.org"
 
-gem 'bookbindery', '10.1.14'
+gem 'faraday', '0.17.4'
```

Comment:
I think a better approach here is simply to specify only one dependency:

gem 'bookbindery'

And let it pick up the rest automatically.  Having a trimmed-down Gemfile with just that line is a best practice for us.

Reply:
Doing a `bundle install` with just that line and no `Gemfile.lock` pulls in an even later version of faraday:

```
$ ag faraday
Gemfile.lock
62:      faraday (~> 1)
71:    faraday (1.4.2)
72:      faraday-em_http (~> 1.0)
73:      faraday-em_synchrony (~> 1.0)
74:      faraday-excon (~> 1.1)
75:      faraday-net_http (~> 1.0)
76:      faraday-net_http_persistent (~> 1.1)
79:    faraday-em_http (1.0.0)
80:    faraday-em_synchrony (1.0.0)
81:    faraday-excon (1.1.0)
82:    faraday-net_http (1.0.1)
83:    faraday-net_http_persistent (1.1.0)
```

Reply:
Have you followed the steps described [here](https://github.com/apache/geode/tree/develop/geode-book#building-the-documentation) to build the Geode documentation? That are the following:

```
$ cd <geode repo>/geode-book
$ bundle install
$ bundle exec bookbinder bind local
$ cd final_app
$ bundle install
$ rackup
```

If I use a Gemfile containing just the bookbindery 10.1.17 dependency and if I follow these steps, I get the error when executing `rackup`.

I have tried again removing `Gemfile.lock` and creating a new rbenv using Ruby 2.5.5 to ensure I dont have any gem installed, and the problem is there:

```
$ rackup
Traceback (most recent call last):
	27: from /home/alb3rtobr/.rbenv/versions/2.5.5/bin/rackup:23:in `<main>'
	26: from /home/alb3rtobr/.rbenv/versions/2.5.5/bin/rackup:23:in `load'
	25: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/bin/rackup:5:in `<top (required)>'
	24: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:170:in `start'
	23: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:313:in `start'
	22: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:381:in `handle_profiling'
	21: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:314:in `block in start'
	20: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:424:in `wrapped_app'
	19: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:251:in `app'
	18: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/server.rb:351:in `build_app_and_options_from_config'
	17: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/builder.rb:40:in `parse_file'
	16: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/builder.rb:59:in `load_file'
	15: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/builder.rb:65:in `new_from_string'
	14: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/builder.rb:65:in `eval'
	13: from /home/alb3rtobr/IdeaProjects/Nordix/geode/geode-book/final_app/config.ru:in `<main>'
	12: from /home/alb3rtobr/IdeaProjects/Nordix/geode/geode-book/final_app/config.ru:in `new'
	11: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/builder.rb:71:in `initialize'
	10: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rack-2.1.4/lib/rack/builder.rb:71:in `instance_eval'
	 9: from /home/alb3rtobr/IdeaProjects/Nordix/geode/geode-book/final_app/config.ru:3:in `block in <main>'
	 8: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	 7: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	 6: from /home/alb3rtobr/IdeaProjects/Nordix/geode/geode-book/final_app/mail_sender.rb:1:in `<top (required)>'
	 5: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:39:in `require'
	 4: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:128:in `rescue in require'
	 3: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems.rb:217:in `try_activate'
	 2: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems.rb:224:in `rescue in try_activate'
	 1: from /home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems/specification.rb:1438:in `activate'
/home/alb3rtobr/.rbenv/versions/2.5.5/lib/ruby/2.5.0/rubygems/specification.rb:2325:in `raise_if_conflicts': Unable to activate sendgrid-ruby-1.1.6, because faraday-1.4.2 conflicts with faraday (~> 0.9) (Gem::ConflictError)

```

Trying to understand the problem, I found what you mention: the new `geode-book/Gemfile.lock` created includes a new `faraday` version (`1.4.2`). I found this is due to `elasticsearch` is also updated from `7.3.0` to `7.13.0`.

This is the diff in the `geode-book/Gemfile.lock` regarding `elasticsearch`:

```
     dotenv (2.7.6)
-    elasticsearch (7.3.0)
-      elasticsearch-api (= 7.3.0)
-      elasticsearch-transport (= 7.3.0)
-    elasticsearch-api (7.3.0)
+    elasticsearch (7.13.0)
+      elasticsearch-api (= 7.13.0)
+      elasticsearch-transport (= 7.13.0)
+    elasticsearch-api (7.13.0)
       multi_json
-    elasticsearch-transport (7.3.0)
-      faraday
+    elasticsearch-transport (7.13.0)
+      faraday (~> 1)
       multi_json
```
There you can see the new `faraday (~> 1)` dependency introduced. That is causing the installation of `faraday 1.4.2`.
I avoided that by adding the depedency on `elastisearch 7.3.0` in the Gemfile, so `elasticsearch-transport` is not updated and the dependency on `faraday` stay the same.

On the other hand, its true that the `final_app/Gemfile.lock` does not have a conflict with `faraday` versions. It appears in three lines:

```
...
elasticsearch-transport (2.0.2)
    faraday
    multi_json
faraday (0.15.4)
    multipart-post (>= 1.2, < 3)
...
sendgrid-ruby (1.1.6)
      faraday (~> 0.9)
      mimemagic
      smtpapi (~> 0.1)
...
``` 
And the `elasticsearch-transport` version here is quite old... So Im not sure why in my case, running `rackup` requires `faraday (~> 1)`, which causes the problem with the `faraday` version required by `sendgrid-ruby`.

BTW, Im using ubuntu and rbenv, in case it helps.

Reply:
@alb3rtobr Mea culpa - I think the README instructions (last edited by me) may be incorrect. My new understanding is that running rackup alone ignores the Gemfire.lock. Once you are in the final_app directory try using the existing Gemfile.lock (don't regenerate it with `bundle install`), but invoke rackup using the bundler, so it will read the Gemfile.lock:
```
$ cd final_app
$ bundle exec rackup
```
This works for me.

Reply:
Thanks @davebarnes97 , mistery solved! I followed your instructions and it also works for me!  Im going to prepare a new commit with the Gemfile and Gemfile.lock files and also with the changes for the README

Reply:
BTW, thanks @dyozie for your comments too!